rllib/models/torch/model.py

```diff
@@ -0,0 +1,183 @@
+import torch
+from torch import nn
+import tree
+
+from ray.rllib.utils.annotations import (
+    DeveloperAPI,
+    override,
+)
+from ray.rllib.models.temp_spec_classes import TensorDict, ModelConfig
+from ray.rllib.models.base_model import RecurrentModel, Model, ModelIO
+
+
+class TorchModelIO(ModelIO):
+    """Save/Load mixin for torch models
+
+    Examples:
+        >>> model.save("/tmp/model_weights.cpt")
+        >>> model.load("/tmp/model_weights.cpt")
+    """
+
+    @DeveloperAPI
+    @override(ModelIO)
+    def save(self, path: str) -> None:
+        """Saves the state dict to the specified path
+
+        Args:
+            path: Path on disk the checkpoint is saved to
+
+        """
+        torch.save(self.state_dict(), path)
+
+    @DeveloperAPI
+    @override(ModelIO)
+    def load(self, path: str) -> RecurrentModel:
+        """Loads the state dict from the specified path
+
+        Args:
+            path: Path on disk to load the checkpoint from
+        """
+        self.load_state_dict(torch.load(path))
+
+
+class TorchRecurrentModel(RecurrentModel, nn.Module, TorchModelIO):
+    """The base class for recurrent pytorch models.
+
+    If implementing a custom recurrent model, you likely want to inherit
+    from this model.
+
+    Args:
+        config: The config used to construct the model
+
+    Required Attributes:
+        input_spec: SpecDict: Denotes the input keys and shapes passed to `unroll`
+        output_spec: SpecDict: Denotes the output keys and shapes returned from
+            `unroll`
+        prev_state_spec: SpecDict: Denotes the keys and shapes for the input
+            recurrent states to the model
+        next_state_spec: SpecDict: Denotes the keys and shapes for the
+            recurrent states output by the model
+
+    Required Overrides:
+        # Define unrolling (forward pass) over a sequence of inputs
+        _unroll(self, inputs: TensorDict, prev_state: TensorDict, **kwargs)
+            -> Tuple[TensorDict, TensorDict]
+
+    Optional Overrides:
+        # Define the initial state, if a zero tensor is insufficient
+        # the returned TensorDict must match the prev_state_spec
+        _initial_state(self) -> TensorDict
+
+        # Save model weights to path
+        save(self, path: str) -> None
+
+        # Load model weights from path
+        load(self, path: str) -> None
+
+    Examples:
+        >>> class MyCustomModel(TorchRecurrentModel):
+        ...     def __init__(self, input_size, output_size, recurrent_size):
+        ...         self.input_spec = SpecDict(
```

Comment:
respect the getters :)

Reply:
also use super(). Make sure in the docs you mention super should be called for init.

Reply:
also for input take config as input. Don't break them up here. We'll figure out the data type for config objects later. For now assume they are dictionaries.

Reply:
fixed